pkg/queue/stats.go

```diff
@@ -41,36 +41,45 @@ const (
 )
 
 // NewStats instantiates a new instance of Stats.
-func NewStats(startedAt time.Time, reqCh chan ReqEvent, reportCh <-chan time.Time, report func(float64, float64, float64, float64)) {
+func NewStats(startedAt time.Time, reqCh chan ReqEvent, reportInterval time.Duration,
+	report func(float64, float64, float64, float64)) {
+	newStatsWithTicker(startedAt, reqCh, reportInterval, time.NewTicker(reportInterval).C, report)
+}
+
+func newStatsWithTicker(startedAt time.Time, reqCh chan ReqEvent, reportInterval time.Duration,
+	ticker <-chan time.Time, report func(float64, float64, float64, float64)) {
+
 	go func() {
 		var (
-			requestCount       float64
-			proxiedCount       float64
-			concurrency        int
-			proxiedConcurrency int
-		)
+			intervalSecs = reportInterval.Seconds()
 
-		lastChange := startedAt
-		timeOnConcurrency := make(map[int]time.Duration)
-		timeOnProxiedConcurrency := make(map[int]time.Duration)
+			// State variables that track the current state. Not resettet after reporting.
+			concurrency, proxiedConcurrency float64
+			lastChange                      = startedAt
+
+			// Reporting variables that track state over the current window. Resettet after
+			// reporting.
+			requestCount, proxiedCount, averageConcurrency, averageProxiedConcurrency float64
+		)
 
 		// Updates the lastChanged/timeOnConcurrency state
 		// Note: due to nature of the channels used below, the ReportChan
 		// can race the ReqChan, thus an event can arrive that has a lower
 		// timestamp than `lastChange`. This is ignored, since it only makes
 		// for very slight differences.
-		updateState := func(concurrency int, time time.Time) {
+		updateState := func(time time.Time) {
 			if durationSinceChange := time.Sub(lastChange); durationSinceChange > 0 {
-				timeOnConcurrency[concurrency] += durationSinceChange
-				timeOnProxiedConcurrency[proxiedConcurrency] += durationSinceChange
+				durationSecs := durationSinceChange.Seconds()
+				averageConcurrency += concurrency * durationSecs / intervalSecs
```

Comment:
division by `intervalSecs` presumes that timer will fire at exact times.
This is what I see in reality
```
### Total: 1.000086638s, MAP: map[0:1.000086638s]
### Total: 999.99774ms, MAP: map[0:38.236077ms 1:351.894744ms 2:278.982325ms 3:275.674963ms 4:55.209631ms]
### Total: 999.99774ms, MAP: map[0:999.99774ms]
### Total: 999.97725ms, MAP: map[0:17.315172ms 1:141.835334ms 2:461.637018ms 3:140.25955ms 4:60.823269ms 5:178.106907ms]
### Total: 999.97725ms, MAP: map[0:999.97725ms]
### Total: 999.985261ms, MAP: map[0:103.191585ms 1:260.756072ms 2:157.057611ms 3:304.220203ms 4:174.407664ms 5:352.126µs]
### Total: 999.985261ms, MAP: map[0:999.985261ms]
### Total: 1.000040401s, MAP: map[0:87.884792ms 1:147.853839ms 2:494.923658ms 3:217.277522ms 4:14.681401ms 5:37.419189ms]
### Total: 1.000040401s, MAP: map[0:1.000040401s]
### Total: 1.000016701s, MAP: map[2:116.952837ms 3:273.266335ms 4:174.320951ms 5:367.449601ms 6:68.026977ms]
### Total: 1.000016701s, MAP: map[0:1.000016701s]
### Total: 999.959674ms, MAP: map[1:136.771914ms 2:231.826001ms 3:308.400128ms 4:270.929664ms 5:49.481667ms 6:2.5503ms]
### Total: 999.959674ms, MAP: map[0:999.959674ms]

```
May be we care, may be we don't, it's all under 1ms :-) but it's less precise than before.

Reply:
Also I think division can be done once when reporting, no? `a/x + b/x + c/x = (a+b+c)/x`

Reply:
I think rather than using constant you can sum the timediffs here and divide by that in the report. So the best of both worlds :D

Reply:
"Less precise" is debatable imo. It's all within rounding errors anyway.

True on doing the revision once!